### PR TITLE
build: Fix a few compilation issues with Clang 7 and -Werror

### PR DESCRIPTION
**Pull request by vasild**

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

**Comment by fanquake** [comment]
@vasild Have you also submitted any of these boost m4 changes upstream?

**Comment by theuni** [comment]
I'd really rather not go down this path. We can't expect to find only warning-free headers, especially with system libs that we can't control.

Have you tried using ```--enable-werror``` rather than adding directly to *FLAGS? That option is there to solve exactly this issue, as well as to whitelist no-false-positive warnings. We could probably stand to be more aggressive with those.

**Comment by practicalswift** [comment]
@theuni But the `net_processing.h` change is OK?

**Comment by vasild** [comment]
@fanquake 
> Have you submitted any of these boost m4 changes upstream?

No. I noticed these files are imported from another project (boost), but saw that there are already some local mods, so decided that it may be ok to change them.

Anyway, I should submit these to boost.

**Comment by vasild** [comment]
@theuni 
> We can't expect to find only warning-free headers, especially with system libs that we can't control.

Yes, that is true. But notice that in this case the warnings do not come from system headers or from headers that are outside of our control. They come from files within the bitcoin codebase: `configure.ac`, `build-aux/m4/ax_boost_chrono.m4` and `build-aux/m4/ax_boost_unit_test_framework.m4`.

If we are more relaxed during configure-time checks (e.g. no -Werror), then there is a chance that we pronounce a library which produces warnings as "available" and later if we build with stricter flags (e.g. -Werror), then the build itself would choke when it tries to use that library.

> Have you tried using --enable-werror rather than adding directly to *FLAGS? That option is there to solve exactly this issue, as well as to whitelist no-false-positive warnings.

Yes, it only uses the stricter flags when compiling files from `src/` and not during the configure checks. I was upset that it only adds `-Werror=vla -Werror=thread-safety-analysis`.

> We could probably stand to be more aggressive with those.

Yes. I would suggest to use just `-Werror` instead of `-Werror=vla -Werror=thread-safety-analysis`. That could make it fail in environments where it previously succeeded. I guess it would need more extensive testing - at least with all supported compilers, right?

Should I adjust this pull request to do that - use just `-Werror`?

Regardless of that, I think the changes to `configure.ac` are good and can/should go in. The changes to `build-aux/` have the issue of modifying imported source from boost and the risk of those changes being wiped away in case of re-import of a newer version (but there is a project to ditch boost altogether).

How to proceed?

File: src/net_processing.h [hunk 1]
```diff
@@ -42,6 +42,8 @@ class PeerLogicValidation : public CValidationInterface, public NetEventsInterfa
 public:
     explicit PeerLogicValidation(CConnman* connman, CScheduler &scheduler);
 
+    virtual ~PeerLogicValidation() = default;
+
```
Comment (laanwj) [inline src/net_processing.h]: ACK, also seen this warning.

**Review by laanwj — COMMENTED** [review]
Defer to @theuni on the build system changes, but adding the virtual destructor when inheriting from a class with virtual methods is good practice.

**Comment by vasild** [comment]
@fanquake 
> Have you submitted any of these boost m4 changes upstream?

There you go: https://github.com/peti/autoconf-archive/pull/148

**Comment by practicalswift** [comment]
What about submitting the `PeerLogicValidation` fix as a separate PR? I've seen that warning myself and it would be nice to get rid of it :-) The m4 changes seem to require more discussion.

**Comment by vasild** [comment]
@practicalswift

> What about submitting the uncontroversial PeerLogicValidation fix as a separate PR? I've seen that warning myself and it would be nice to get rid of it :-)

there you go: https://github.com/bitcoin/bitcoin/pull/12680

> The m4 changes seem to require more discussion.

Yes. The m4 boost changes have been taken upstream, so I have changed this pull request to apply/take all changes from upstream to those two files. I guess those should now be ok, because it just updates from upstream, @theuni?

If the two changes in `configure.ac` are undesirable (`MSG_*` and `libminiupnpc` check) I could ditch them away from this pull request.

Unrelated to the outcome of the above, I will consider changing `--enable-werror` to use just `-Werror` instead of `-Werror=vla -Werror=thread-safety-analysis`.

Cheers!

File: configure.ac [hunk 2]
```diff
@@ -771,7 +771,7 @@ dnl Check for libminiupnpc (optional)
 if test x$use_upnp != xno; then
   AC_CHECK_HEADERS(
     [miniupnpc/miniwget.h miniupnpc/miniupnpc.h miniupnpc/upnpcommands.h miniupnpc/upnperrors.h],
-    [AC_CHECK_LIB([miniupnpc], [main],[MINIUPNPC_LIBS=-lminiupnpc], [have_miniupnpc=no])],
+    [AC_CHECK_LIB([miniupnpc], [upnpDiscover], [MINIUPNPC_LIBS=-lminiupnpc], [have_miniupnpc=no])],
```
Comment (eklitzke) [inline configure.ac]: I don't see how this is related to Clang 7, this is checking for the visibility of a symbol. This might be needed on your version of upnp but if so that should be another PR in my opinion.

File: configure.ac [hunk 3]
```diff
@@ -646,15 +646,15 @@ AC_CHECK_DECLS([__builtin_clz, __builtin_clzl, __builtin_clzll])
 dnl Check for MSG_NOSIGNAL
 AC_MSG_CHECKING(for MSG_NOSIGNAL)
 AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <sys/socket.h>]],
- [[ int f = MSG_NOSIGNAL; ]])],
+ [[ return MSG_NOSIGNAL; ]])],
```
Comment (eklitzke) [inline configure.ac]: This seems broken to me in the first place, isn't `MSG_NOSIGNAL` a preprocessor define? If so there's no need to check for it in configure.ac, we can just use `#if defined()` in `net.cpp` and `netbase.cpp`.

**Review by eklitzke — CHANGES_REQUESTED** [review]

File: configure.ac (hunk 2 above)
Comment (vasild) [reply configure.ac]: The commit message of that change reads:

```
    Do not check for main() in libminiupnpc
    
    main() { main(); } causes "infinite recursion" compilation warning
    which with -Werror fails the check.
```

Maybe that is not clear enough, let me elaborate: `AC_CHECK_LIB([lib], [func], ...)` generates a test program like `main() { func(); ... }`  and tries to link it with `-llib`. The idea being that if `func()` is not present in `lib` it will fail to link with an undefined reference error. So the current usage leads to a test program like `main() { main(); ... }` which generates a legit compiler warning of "infinite recursion". At least Clang 7 emits that warning, maybe other compilers do so too.

Anyway, I think checking for function `main()` in `libminiupnpc` is a misuse of `AC_CHECK_LIB` because `main()` does not exist in `libminiupnpc` and it passes because it relies on how the check is performed internally. If `AC_CHECK_LIB` starts using another method of checking if a function is present in a library (e.g. `nm(1)`) then the check may as well fail.

File: configure.ac (hunk 3 above)
Comment (vasild) [reply configure.ac]: `MSG_NOSIGNAL` is a macro in my environment and maybe in lots of other environments too. So this check can be wiped away from `configure.ac` and `#ifdef MSG_NOSIGNAL` be used instead of `#ifdef HAVE_MSG_NOSIGNAL` in the source code. On my environment that will work.

I guess the reason for this check in `configure.ac` is that we are not guaranteed that it will be a macro. Maybe somewhere it could be `const int MSG_NOSIGNAL = ...`?

My change intends to fix an "unused variable" warning, with a minimum chance of breaking something else.

Comment (eklitzke) [reply configure.ac]: I believe all of these options are required to be macros by the standard (precisely for reasons like this). Also note that `const int FOO = 1` and `#define FOO 1` are not the same, using a const value is strictly worse as it requires declaring an extern value that gets storage space allocated in the program. So there's no reason an implementation would ever do things that way.

The problem with keeping logic like this "just in case" is that these `AC_COMPILE_IFELSE` checks are relatively expensive, so we should only be using them for features that actually require compilation.

File: configure.ac (hunk 2 above)
Comment (eklitzke) [reply configure.ac]: That makes sense, thanks for clarifying.

File: configure.ac (hunk 3 above)
Comment (vasild) [reply configure.ac]: You are right - `const int` requires storage somewhere. I have removed the checks from `configure.ac` and updated the pull request.

**Review by theuni — APPROVED** [review]
utACK 8ae413235d4b3f68f0858a3ce2cd65291f9cbe64. Thanks!

**Comment by eklitzke** [comment]
utACK 8ae413235d4b3f68f0858a3ce2cd65291f9cbe64
